### ISSUE TITLE
Set simplification of `|Set` confuses set contents

### ISSUE THREAD

**Issue by jberthold** [runtimeverification/haskell-backend#3301]
The following definition (with a simple simplification rule involving set matching) produces wrong simplifications and bogus side conditions:

`test.k`
```
module TEST
  imports SET
  imports BOOL

  syntax Thing ::= "a" | "b"

  rule (S SetItem(X)) |Set S => S SetItem(X) ensures notBool (X in S) [simplification]

  configuration <k> $PGM:Test </k>

  syntax Test ::= union ( Set ) | result ( Set )
  rule <k> union(E) => result(E) </k>

endmodule
```
`spec.k`
```
module SPEC
  imports TEST

  claim [union]: <k> union ( (SetItem(_X) SetItem(a)) |Set SetItem(b)) => . </k>
endmodule
```
The state dump for this "proof" contains a wrong side condition introduced by the simplification rule.

```
$ kprove --claims SPEC.union spec.k
kore-exec: [129231] Warning (WarnStuckClaimState):
    The configuration's term doesn't unify with the destination's term and the configuration cannot be rewritten further. Location: /home/jost/work/RV/code/playground/set-simplification-confusing-variables2/spec.k:4:18-4:81
Context:
    (InfoReachability) while checking the implication
  #Not ( {
    _X
  #Equals
    b
  } )
#And
  <k>
    result ( SetItem ( _X )
    SetItem ( b ) ) ~> .
  </k>
#And
  {
    false
  #Equals
    _X in SetItem ( b )
  }
[Error] Prover: backend terminated because the configuration cannot be
rewritten further. See output for more details.
```
**Note the result without `{a}` and the bogus side condition `! ( _X ∈ { b }` at the bottom**,  which I can only explain by assuming that the `S` in the rule is matched with both `{a}` and `{b}` (dropping `{a}` later) and the simplification is (wrongly) applied.

This is likely to be related to #3294 as it has the same matching pattern.

**Comment by jberthold** [comment]
This bug is also observable when flipping the arguments to `|Set`, i.e. with a rule 
```
  rule S |Set (S SetItem(X)) => S SetItem(X)
                    ensures notBool (X in S) [simplification]
```
and a claim
```
claim [union2]: <k> union ( SetItem(a) |Set (SetItem(_X) SetItem(b))) => . </k>
```
Result is
```
  #Not ( {
    _X
  #Equals
    a
  } )
#And
  <k>
    result ( SetItem ( _X )
    SetItem ( a ) ) ~> .
  </k>
#And
  {
    false
  #Equals
    _X in SetItem ( a )
  }
```
inventing the `! (_X == a)`

**Comment by ana-pantilie** [comment]
This should be fixed by doing https://github.com/runtimeverification/haskell-backend/issues/2739

**Comment by jberthold** [comment]
Fixed in #3334 (with a related integration test), closing this as duplicate
